Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -24,30 +30,145 @@ public class DafnyHoverHandler : HoverHandlerBase {
       };
     }
 
-    public async override Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
+    public override async Task<Hover?> Handle(HoverParams request, CancellationToken cancellationToken) {
       logger.LogTrace("received hover request for {Document}", request.TextDocument);
       var document = await documents.GetDocumentAsync(request.TextDocument);
       if (document == null) {
         logger.LogWarning("the document {Document} is not loaded", request.TextDocument);
         return null;
       }
+      var diagnosticHoverContent = GetDiagnosticsHover(document, request.Position);
       if (!document.SymbolTable.TryGetSymbolAt(request.Position, out var symbol)) {
         logger.LogDebug("no symbol was found at {Position} in {Document}", request.Position, request.TextDocument);
+      }
+
+      var symbolHoverContent = symbol != null ? CreateSymbolMarkdown(symbol, cancellationToken) : null;
+      if (diagnosticHoverContent == null && symbolHoverContent == null) {
         return null;
       }
-      return CreateHover(symbol, cancellationToken);
+
+      var hoverContent = diagnosticHoverContent == null ? "" : diagnosticHoverContent;
+      hoverContent += symbolHoverContent == null ? (hoverContent != "" ? "  \n" : "") : symbolHoverContent;
+      return CreateMarkdownHover(hoverContent);
+    }
+
+    private string? GetDiagnosticsHover(DafnyDocument document, Position position) {
```

Comment:
If you review this PR linearly, you might not yet be familiar of the new NodeDiagnostic data structure that is both being sent to the client and used in the server to report verification diagnostics on hover.

Roughly, everything ending with `NodeDiagnostic` inherits from this base class, and although they have `Children` of type `NodeDiagnostic `they are really"nested" in a tree like the following
```
DocumentNodeDiagnostic {
  Children: List<TopLevelDeclMemberNodeDiagnostic>
}
TopLevelDeclMemberNodeDiagnostic {
  Children: List<ImplementationNodeDiagnostic>
  AssertionBatches: List<AssertionBatchNodeDiagnostic>
}

ImplementationNodeDiagnostic {
  Children: List<AssertionNodeDiagnostic>
}
AssertionBatchNodeDiagnostic {
  Children: List<AssertionNodeDiagnostic>
}
```

The idea is that `AssertionBatchNodeDiagnostic ` are automatically computed from all the `ImplementationNodeDiagnostic`, and they are used for all the reporting, because users don't care about implementations.
However, we do care about implementations to be able to relocate previous diagnostics. Hence the tree structure.